File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/06ef699c0a10b4f1f8a9d78578e48a08.go
package main

import (
	"fmt"
)

type Matrix [][]int

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func solve(a Matrix, w []int, dp []int, l, r int) {
	if r-l == 1 {
		sum := 0
		for i := 0; i <= len(a[l]); i++ {
			ans = max(ans, dp[k-i]+sum)
			if i != len(a[l]) {
				sum += a[l][i]
			}
		}
		return
	}
	m := (r + l) / 2
	dp2 := make([]int, k+1)
	copy(dp2, dp)
	for i := m; i < r; i++ {
		for j := k - 1; j >= 0; j-- {
			if len(a[i])+j <= k {
				dp2[j+len(a[i])] = max(dp2[j+len(a[i])], dp2[j]+w[i])
			}
		}
	}
	solve(a, w, dp2, l, m)
	for i := l; i < m; i++ {
		for j := k - 1; j >= 0; j-- {
			if len(a[i])+j <= k {
				dp[j+len(a[i])] = max(dp[j+len(a[i])], dp[j]+w[i])
			}
		}
	}
	solve(a, w, dp, m, r)
}

func main() {
	var n, k int
	fmt.Scan(&n, &k)
	a := make(Matrix, n)
	w := make([]int, n)
	for i := 0; i < n; i++ {
		var t int
		fmt.Scan(&t)
		for j := 0; j < t; j++ {
			var tmp int
			fmt.Scan(&tmp)
			if len(a[i]) < k {
				a[i] = append(a[i], tmp)
			}
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < len(a[i]); j++ {
			w[i] += a[i][j]
		}
	}
	ans := 0
	dp := make([]int, k+1)
	solve(a, w, dp, 0, n)
	fmt.Println(ans)
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/06ef699c0a10b4f1f8a9d78578e48a08.py
import sys
from typing import List, Tuple

MOD = 1000000007

def solve(n: int, k: int, a: List[List[int]]) -> int:
    def dfs(l: int, r: int, dp: List[int]) -> None:
        if r - l == 1:
            sum_val = 0
            for i in range(len(a[l]) + 1):
                ans[0] = max(ans[0], dp[k - i] + sum_val)
                if i != len(a[l]):
                    sum_val += a[l][i]
            return
        m = (l + r) // 2
        dp2 = dp.copy()
        for i in range(m, r):
            for j in range(k - 1, -1, -1):
                if len(a[i]) + j <= k:
                    dp2[j + len(a[i])] = max(dp2[j + len(a[i])], dp2[j] + sum(a[i]))
        dfs(l, m, dp2)
        for i in range(l, m):
            for j in range(k - 1, -1, -1):
                if len(a[i]) + j <= k:
                    dp[j + len(a[i])] = max(dp[j + len(a[i])], dp[j] + sum(a[i]))
        dfs(m, r, dp)
    
    ans = [0]
    dp = [0] * (k + 1)
    dfs(0, n, dp)
    return ans[0]

def from_input_string(input_string: str) -> Tuple[int, int, List[List[int]]]:
    lines = input_string.strip().split('\n')
    n, k = map(int, lines[0].strip().split())
    a = [list(map(int, x.strip().split())) for x in lines[1:]]
    return n, k, a

def to_input_string(inputs: Tuple[int, int, List[List[int]]]) -> str:
    n, k, a = inputs
    res = []
    res.append(f"{n} {k}")
    for row in a:
        res.append(' '.join(str(x) for x in row))
    return '\n'.join(res)

def from_output_string(output_string: str) -> int:
    return int(output_string.strip())

def to_output_string(output: int) -> str:
    return str(output)
